File: src/libs/ThirdParty.VBSuppliers/Properties/Stuba/Stuba.vb
Imports System.Xml
Imports Intuitive.Net.WebRequests
Imports Intuitive.Functions
Imports ThirdParty.Abstractions
Imports ThirdParty.Abstractions.Models.Property.Booking
Imports ThirdParty.Abstractions.Models
Imports Intuitive.Helpers.Extensions
Imports ThirdParty.Abstractions.Constants
Imports ThirdParty.Abstractions.Lookups
Imports Intuitive
Imports System.Net.Http

Public Class Stuba
    Implements IThirdParty

    Private ReadOnly _settings As IStubaSettings

    Private ReadOnly _support As ITPSupport

    Private ReadOnly _httpClient As HttpClient

    Public Sub New(settings As IStubaSettings, support As ITPSupport, httpClient As HttpClient)
        _settings = Ensure.IsNotNull(settings, NameOf(settings))
        _support = Ensure.IsNotNull(support, NameOf(support))
        _httpClient = Ensure.IsNotNull(httpClient, NameOf(httpClient))
    End Sub

    Public Function PreBook(oPropertyDetails As PropertyDetails) As Boolean Implements IThirdParty.PreBook
        Dim bReturn As Boolean

        Try
            Dim oXML As XmlDocument = SendRequest("Pre-Book", BookReservationRequest(oPropertyDetails, False), oPropertyDetails)
            ExtractErrata(oXML, oPropertyDetails)
            bReturn = CostsAndCancellation(oPropertyDetails, oXML)
        Catch ex As Exception
            oPropertyDetails.Warnings.AddNew("PreBookException", ex.ToString)
            bReturn = False
        End Try
        Return bReturn
    End Function

    Public Function Book(oPropertyDetails As PropertyDetails) As String Implements IThirdParty.Book
        Dim sReference As String = ""

        Try
            Dim oXML As XmlDocument = SendRequest("Book", BookReservationRequest(oPropertyDetails, True), oPropertyDetails)

            If oXML.SelectSingleNode("BookingCreateResult/Booking/HotelBooking/Status").InnerText.ToLower = "confirmed" Then
                sReference = oXML.SelectSingleNode("BookingCreateResult/Booking/Id").InnerText
            End If
        Catch ex As Exception
            oPropertyDetails.Warnings.AddNew("BookException", ex.ToString)
            sReference = "Failed"
        End Try
        Return sReference
    End Function

    Public Function CancelBooking(oPropertyDetails As PropertyDetails) As ThirdPartyCancellationResponse Implements IThirdParty.CancelBooking
        Dim oThirdPartyCancellationResponse As New ThirdPartyCancellationResponse

        Try
            Dim oXML As XmlDocument = SendRequest("Cancel", CancelRequest(oPropertyDetails, True), oPropertyDetails)

            If oXML.SelectSingleNode("BookingCancelResult/Booking/HotelBooking/Status").InnerText.ToLower = "cancelled" Then
                oThirdPartyCancellationResponse.TPCancellationReference = oXML.SelectSingleNode("BookingCancelResult/Booking/HotelBooking/Id").InnerText
                oThirdPartyCancellationResponse.Success = True
            End If

        Catch ex As Exception
            oPropertyDetails.Warnings.AddNew("CancelException", ex.ToString)
            oThirdPartyCancellationResponse.TPCancellationReference = "Failed"
        End Try
        Return oThirdPartyCancellationResponse
    End Function

    Public Function GetCancellationCost(oPropertyDetails As PropertyDetails) As ThirdPartyCancellationFeeResult Implements IThirdParty.GetCancellationCost
        Dim oThirdPartyCancellationFeeResult As New ThirdPartyCancellationFeeResult

        Try
            Dim oXML As XmlDocument = SendRequest("Cancellation Charge", CancelRequest(oPropertyDetails, False), oPropertyDetails)
            Dim nTotalCancellationFee As Decimal = 0

            For Each oRoom As RoomDetails In oPropertyDetails.Rooms
                Dim dLastDate As New Date(1900, 1, 1)
                Dim nRoomAmount As Decimal = 0
                Dim oRoomNode As XmlNode = oXML.SelectSingleNode(String.Format("BookingCancelResult/Booking/HotelBooking/Room[RoomType/@code = '{0}']", oRoom.RoomTypeCode))

                For Each oCXL As XmlNode In oRoomNode.SelectNodes("CanxFees/Fee")
                    Dim dStartDate As Date = oCXL.SelectSingleNode("@from").InnerText.ToSafeDate
                    If dStartDate < Date.Now AndAlso dStartDate > dLastDate Then
                        nRoomAmount = oCXL.SelectSingleNode("Amount/@amt").InnerText.ToSafeDecimal()
                    End If
                Next
                nTotalCancellationFee += nRoomAmount
            Next

            oThirdPartyCancellationFeeResult.Amount = nTotalCancellationFee
            oThirdPartyCancellationFeeResult.CurrencyCode = oPropertyDetails.CurrencyCode

        Catch ex As Exception
            oPropertyDetails.Warnings.AddNew("CancellationChargeException", ex.ToString)
        End Try

        Return oThirdPartyCancellationFeeResult
    End Function

    Private Function SendRequest(ByVal RequestType As String, XML As String, ByVal PropertyDetails As PropertyDetails) As XmlDocument

        Dim oRequest As New Request
        With oRequest
            .EndPoint = _settings.URL(PropertyDetails)
            .SetRequest(XML)
            .Method = eRequestMethod.POST
            .Source = ThirdParties.STUBA
            .LogFileName = RequestType
            .CreateLog = True
            .UseGZip = True
            .Send(_httpClient)
        End With

        If oRequest.RequestLog <> "" Then
            PropertyDetails.Logs.AddNew(ThirdParties.STUBA, "Stuba " & RequestType & " Request", oRequest.RequestLog)
        End If

        If oRequest.ResponseLog <> "" Then
            PropertyDetails.Logs.AddNew(ThirdParties.STUBA, "Stuba " & RequestType & " Response", oRequest.ResponseLog)
        End If

        Return oRequest.ResponseXML
    End Function

    Private Function BookReservationRequest(ByVal oPropertyDetails As PropertyDetails, ByVal Confirm As Boolean) As String

        Dim sOrg As String = _settings.Organisation(oPropertyDetails)
        Dim sUser As String = _settings.Username(oPropertyDetails)
        Dim sPassword As String = _settings.Password(oPropertyDetails)
        Dim sVersion As String = _settings.Version(oPropertyDetails)
        Dim sCurrencyCode As String = _settings.Currency(oPropertyDetails)

        EnsurePassengerNamesNotNull(oPropertyDetails)
        Dim oRequest As XElement =
            <BookingCreate>
                <Authority>
                    <Org><%= sOrg %></Org>
                    <User><%= sUser %></User>
                    <Password><%= sPassword %></Password>
                    <Currency><%= sCurrencyCode %></Currency>
                    <Version><%= sVersion %></Version>
                </Authority>
                <HotelBooking>
                    <QuoteId><%= oPropertyDetails.Rooms.First().ThirdPartyReference.Split("|"c)(1) %></QuoteId>
                    <HotelStayDetails>
                        <Nationality><%= GetNationality(oPropertyDetails) %></Nationality>
                        <%=
                            From oRoom As RoomDetails In oPropertyDetails.Rooms
                            Select <Room>
                                       <Guests>
                                           <%=
                                               From oGuest As Passenger In oRoom.Passengers.Where(Function(p) p.PassengerType = PassengerType.Adult)
                                               Select <Adult title=<%= oGuest.Title %> first=<%= oGuest.FirstName %> last=<%= oGuest.LastName %>/>
                                           %>
                                           <%=
                                               From oGuest As Passenger In oRoom.Passengers.Where(Function(p) p.PassengerType <> PassengerType.Adult)
                                               Select <Child title=<%= oGuest.Title %> first=<%= oGuest.FirstName %> last=<%= oGuest.LastName %> age=<%= oGuest.Age %>/>
                                           %>
                                       </Guests>
                                   </Room>
                        %>
                    </HotelStayDetails>
                </HotelBooking>
                <CommitLevel><%= If(Confirm, "confirm", "prepare") %></CommitLevel>
            </BookingCreate>

        Return oRequest.ToString
    End Function

    Private Function CancelRequest(ByVal oPropertyDetails As PropertyDetails, ByVal bConfirm As Boolean) As String
        Dim sOrg As String = _settings.Organisation(oPropertyDetails)
        Dim sUser As String = _settings.Username(oPropertyDetails)
        Dim sPassword As String = _settings.Password(oPropertyDetails)
        Dim sVersion As String = _settings.Version(oPropertyDetails)
        Dim sCurrencyCode As String = _settings.Currency(oPropertyDetails)

        Dim oRequest As XElement =
            <BookingCancel>
                <Authority>
                    <Org><%= sOrg %></Org>
                    <User><%= sUser %></User>
                    <Password><%= sPassword %></Password>
                    <Currency><%= sCurrencyCode %></Currency>
                    <Version><%= sVersion %></Version>
                </Authority>
                <BookingId><%= oPropertyDetails.SourceReference %></BookingId>
                <CommitLevel><%= If(bConfirm, "confirm", "prepare") %></CommitLevel>
            </BookingCancel>

        Return oRequest.ToString()
    End Function

    Private Sub EnsurePassengerNamesNotNull(propertyDetails As PropertyDetails)
        Dim rand As New Random
        Dim charAsInt As Integer = Char.ConvertToUtf32("A", 0)
        For Each guest As Passenger In propertyDetails.Rooms.SelectMany(Function(room) room.Passengers)
            If String.IsNullOrEmpty(guest.Title) Then
#Disable Warning SCS0005 ' Weak random generator
                guest.Title = If(rand.Next(2) = 0, "Ms", "Mr")
#Enable Warning SCS0005 ' Weak random generator
            End If
            If String.IsNullOrEmpty(guest.FirstName) Then
                guest.FirstName = Char.ConvertFromUtf32(charAsInt)
                charAsInt += 1 'names need to be unique
            End If
            If String.IsNullOrEmpty(guest.LastName) Then
                guest.LastName = "Test"
            End If
        Next
    End Sub

    Private Function GetNationality(propertyDetails As PropertyDetails) As String
        Dim sNationality As String = ""
        If propertyDetails.LeadGuestNationalityID <> 0 Then
            sNationality = _support.TPNationalityLookup(ThirdParties.STUBA, propertyDetails.LeadGuestNationalityID)
        End If
        If sNationality = "" Then
            sNationality = _settings.Nationality(propertyDetails)
        End If
        Return sNationality
    End Function

    Private Function CostsAndCancellation(ByVal oPropertyDetails As PropertyDetails, ByVal oXML As XmlDocument) As Boolean
        Dim bAvailable As Boolean = False

        For Each oRoom As RoomDetails In oPropertyDetails.Rooms

            'Check costs
            Dim oRoomNode As XmlNode = oXML.SelectSingleNode(String.Format("BookingCreateResult/Booking/HotelBooking/Room[RoomType/@code = '{0}']", oRoom.RoomTypeCode))
            oRoom.LocalCost = oRoomNode.SelectSingleNode("TotalSellingPrice/@amt").InnerText.ToSafeDecimal()
            oRoom.GrossCost = oRoomNode.SelectSingleNode("TotalSellingPrice/@amt").InnerText.ToSafeDecimal()

            'Cancellation charges
            For Each oCXL As XmlNode In oRoomNode.SelectNodes("CanxFees/Fee")
                Dim dStartDate As Date

                If oCXL.SelectSingleNode("@from") Is Nothing Then
                    dStartDate = Date.Now.Date
                Else
                    dStartDate = oCXL.SelectSingleNode("@from").InnerText.ToSafeDate
                End If

                Dim dEndDate As New Date(2099, 12, 25)

                Dim oNextStartDate As XmlNode = oCXL.NextSibling
                If oNextStartDate IsNot Nothing Then
                    dEndDate = oNextStartDate.SelectSingleNode("@from").InnerText.ToSafeDate.AddDays(-1)
                End If

                Dim nAmount As Decimal = oCXL.SelectSingleNode("Amount/@amt").InnerText.ToSafeDecimal()
                oPropertyDetails.Cancellations.AddNew(dStartDate, dEndDate, nAmount)
            Next

        Next

        oPropertyDetails.LocalCost = oPropertyDetails.Rooms.Sum(Function(r) r.LocalCost)

        oPropertyDetails.Cancellations.Solidify(SolidifyType.Sum)
        bAvailable = (oXML.SelectNodes("BookingCreateResult/Booking/HotelBooking/Room").Count = oPropertyDetails.Rooms.Count)

        Return bAvailable
    End Function

    Private Sub ExtractErrata(oResponse As XmlDocument, oPropertyDetails As PropertyDetails)
        Dim oErrataNodes As XmlNodeList = oResponse.SelectNodes("BookingCreateResult/Booking/HotelBooking/Room/Messages/Message[Type = 'Supplier Notes']/Text")
        If oErrataNodes?.Count > 0 Then
            For Each oErratum As XmlNode In oErrataNodes
                oPropertyDetails.Errata.AddNew("Important Informations", oErratum.InnerText)
            Next
        End If
    End Sub

    Private Function IThirdParty_SupportsLiveCancellation(searchDetails As IThirdPartyAttributeSearch, source As String) As Boolean Implements IThirdParty.SupportsLiveCancellation
        Return _settings.AllowCancellations(searchDetails)
    End Function

#Region "Stuff"

    Public ReadOnly Property SupportsRemarks As Boolean Implements IThirdParty.SupportsRemarks
        Get
            Return False
        End Get
    End Property

    Public ReadOnly Property SupportsBookingSearch As Boolean Implements IThirdParty.SupportsBookingSearch
        Get
            Return False
        End Get
    End Property

    Private Function IThirdParty_TakeSavingFromCommissionMargin(searchDetails As IThirdPartyAttributeSearch) As Boolean Implements IThirdParty.TakeSavingFromCommissionMargin
        Return False
    End Function

    Private Function IThirdParty_OffsetCancellationDays(searchDetails As IThirdPartyAttributeSearch) As Integer Implements IThirdParty.OffsetCancellationDays
        Return _settings.OffsetCancellationDays(searchDetails, False)
    End Function

    Public Function RequiresVCard(info As VirtualCardInfo) As Boolean Implements IThirdParty.RequiresVCard
        Return False
    End Function

    Public Function BookingSearch(oBookingSearchDetails As BookingSearchDetails) As ThirdPartyBookingSearchResults Implements IThirdParty.BookingSearch
        Return New ThirdPartyBookingSearchResults
    End Function

    Public Function BookingStatusUpdate(oPropertyDetails As PropertyDetails) As ThirdPartyBookingStatusUpdateResult Implements IThirdParty.BookingStatusUpdate
        Return New ThirdPartyBookingStatusUpdateResult
    End Function

    Public Function CreateReconciliationReference(sInputReference As String) As String Implements IThirdParty.CreateReconciliationReference
        Return ""
    End Function

    Public Sub EndSession(ByVal oPropertyDetails As PropertyDetails) Implements IThirdParty.EndSession
    End Sub

#End Region

End Class

File: src/libs/ThirdParty.CSSuppliers/Stuba/Stuba.cs
namespace ThirdParty.CSSuppliers.Stuba
{
    using System;
    using System.Linq;
    using System.Net.Http;
    using System.Threading.Tasks;
    using System.Xml;
    using System.Xml.Linq;
    using Intuitive;
    using Intuitive.Helpers.Extensions;
    using Intuitive.Helpers.Net;
    using Microsoft.Extensions.Logging;
    using ThirdParty.Constants;
    using ThirdParty.Interfaces;
    using ThirdParty.Lookups;
    using ThirdParty.Models;
    using ThirdParty.Models.Property.Booking;

    public class Stuba : IThirdParty, ISingleSource
    {
        private readonly IStubaSettings _settings;

        private readonly ITPSupport _support;

        private readonly HttpClient _httpClient;

        private readonly ILogger<Stuba> _logger;

        public string Source => ThirdParties.STUBA;

        public Stuba(IStubaSettings settings, ITPSupport support, HttpClient httpClient, ILogger<Stuba> logger)
        {
            _settings = Ensure.IsNotNull(settings, nameof(settings));
            _support = Ensure.IsNotNull(support, nameof(support));
            _httpClient = Ensure.IsNotNull(httpClient, nameof(httpClient));
            _logger = Ensure.IsNotNull(logger, nameof(logger));
        }

        public async Task<bool> PreBookAsync(PropertyDetails propertyDetails)
        {
            bool result;

            try
            {
                var xml = await SendRequestAsync("Pre-Book", await BookReservationRequestAsync(propertyDetails, false), propertyDetails);
                ExtractErrata(xml, propertyDetails);
                result = CostsAndCancellation(propertyDetails, xml);
            }
            catch (Exception ex)
            {
                propertyDetails.Warnings.AddNew("PreBookException", ex.ToString());
                result = false;
            }
            return result;
        }

        public async Task<string> BookAsync(PropertyDetails propertyDetails)
        {
            string reference = "";

            try
            {
                var xml = await SendRequestAsync("Book", await BookReservationRequestAsync(propertyDetails, true), propertyDetails);

                if (xml.SelectSingleNode("BookingCreateResult/Booking/HotelBooking/Status").InnerText.ToLower() == "confirmed")
                {
                    reference = xml.SelectSingleNode("BookingCreateResult/Booking/Id").InnerText;
                }
            }
            catch (Exception ex)
            {
                propertyDetails.Warnings.AddNew("BookException", ex.ToString());
                reference = "Failed";
            }
            return reference;
        }

        public async Task<ThirdPartyCancellationResponse> CancelBookingAsync(PropertyDetails propertyDetails)
        {
            var thirdPartyCancellationResponse = new ThirdPartyCancellationResponse();

            try
            {
                var xml = await SendRequestAsync("Cancel", CancelRequest(propertyDetails, true), propertyDetails);

                if (xml.SelectSingleNode("BookingCancelResult/Booking/HotelBooking/Status").InnerText.ToLower() == "cancelled")
                {
                    thirdPartyCancellationResponse.TPCancellationReference = xml.SelectSingleNode("BookingCancelResult/Booking/HotelBooking/Id").InnerText;
                    thirdPartyCancellationResponse.Success = true;
                }
            }

            catch (Exception ex)
            {
                propertyDetails.Warnings.AddNew("CancelException", ex.ToString());
                thirdPartyCancellationResponse.TPCancellationReference = "Failed";
            }
            return thirdPartyCancellationResponse;
        }

        public async Task<ThirdPartyCancellationFeeResult> GetCancellationCostAsync(PropertyDetails propertyDetails)
        {
            var thirdPartyCancellationFeeResult = new ThirdPartyCancellationFeeResult();

            try
            {
                var xml = await SendRequestAsync("Cancellation Charge", CancelRequest(propertyDetails, false), propertyDetails);
                decimal totalCancellationFee = 0m;

                foreach (var room in propertyDetails.Rooms)
                {
                    var lastDate = new DateTime(1900, 1, 1);
                    decimal roomAmount = 0m;
                    var roomNode = xml.SelectSingleNode(string.Format("BookingCancelResult/Booking/HotelBooking/Room[RoomType/@code = '{0}']", room.RoomTypeCode));

                    foreach (XmlNode cancelNode in roomNode.SelectNodes("CanxFees/Fee"))
                    {
                        var dStartDate = cancelNode.SelectSingleNode("@from").InnerText.ToSafeDate();
                        if (dStartDate < DateTime.Now && dStartDate > lastDate)
                        {
                            roomAmount = cancelNode.SelectSingleNode("Amount/@amt").InnerText.ToSafeDecimal();
                        }
                    }
                    totalCancellationFee += roomAmount;
                }

                thirdPartyCancellationFeeResult.Amount = totalCancellationFee;
                thirdPartyCancellationFeeResult.CurrencyCode = propertyDetails.ISOCurrencyCode;
            }
            catch (Exception ex)
            {
                propertyDetails.Warnings.AddNew("CancellationChargeException", ex.ToString());
            }

            return thirdPartyCancellationFeeResult;
        }

        private async Task<XmlDocument> SendRequestAsync(string requestType, string xml, PropertyDetails propertyDetails)
        {
            var request = new Request
            {
                EndPoint = _settings.GenericURL(propertyDetails),
                Method = RequestMethod.POST,
                Source = ThirdParties.STUBA,
                LogFileName = requestType,
                CreateLog = true,
                UseGZip = true
            };
            request.SetRequest(xml);

            try
            {
                await request.Send(_httpClient, _logger);
            }
            finally
            {
                propertyDetails.AddLog(requestType, request);
            }

            return request.ResponseXML;
        }

        private async Task<string> BookReservationRequestAsync(PropertyDetails propertyDetails, bool confirm)
        {
            string org = _settings.Organisation(propertyDetails);
            string user = _settings.User(propertyDetails);
            string password = _settings.Password(propertyDetails);
            string version = _settings.Version(propertyDetails);
            string currencyCode = _settings.Currency(propertyDetails);

            EnsurePassengerNamesNotNull(propertyDetails);
            var request = new XElement("BookingCreate",
                            new XElement("Authority",
                                new XElement("Org", org),
                                new XElement("User", user),
                                new XElement("Password", password),
                                new XElement("Currency", currencyCode),
                                new XElement("Version", version)
                            ),
                            new XElement("HotelBooking",
                            new XElement("QuoteId", propertyDetails.Rooms.FirstOrDefault().ThirdPartyReference.Split('|')[1].ToSafeString()),
                            new XElement("HotelStayDetails",
                            new XElement("Nationality", await GetNationalityAsync(propertyDetails)), from oRoom in propertyDetails.Rooms
                                select new XElement("Room",
                                    new XElement("Guests", from oGuest in oRoom.Passengers.Where(p => p.PassengerType == PassengerType.Adult)
                                        select new XElement("Adult", new XAttribute("title", oGuest.Title), new XAttribute("first", oGuest.FirstName), new XAttribute("last", oGuest.LastName)), from oGuest in oRoom.Passengers.Where(p => p.PassengerType != PassengerType.Adult)
                                        select new XElement("Child", new XAttribute("title", oGuest.Title), new XAttribute("first", oGuest.FirstName), new XAttribute("last", oGuest.LastName), new XAttribute("age", oGuest.Age)))))),
                            new XElement("CommitLevel", confirm ? "confirm" : "prepare")
                            );
            return request.ToString();
        }

        private string CancelRequest(PropertyDetails propertyDetails, bool confirm)
        {
            string org = _settings.Organisation(propertyDetails);
            string user = _settings.User(propertyDetails);
            string password = _settings.Password(propertyDetails);
            string version = _settings.Version(propertyDetails);
            string currencyCode = _settings.Currency(propertyDetails);

            var request = new XElement("BookingCancel",
                            new XElement("Authority",
                            new XElement("Org", org),
                            new XElement("User", user),
                            new XElement("Password", password),
                            new XElement("Currency", currencyCode),
                            new XElement("Version", version)),
                                new XElement("BookingId", propertyDetails.SourceReference),
                                new XElement("CommitLevel", confirm ? "confirm" : "prepare")
                            );

            return request.ToString();
        }

        private void EnsurePassengerNamesNotNull(PropertyDetails propertyDetails)
        {
            var rand = new Random();
            int charAsInt = char.ConvertToUtf32("A", 0);
            foreach (var guest in propertyDetails.Rooms.SelectMany(room => room.Passengers))
            {
                if (string.IsNullOrEmpty(guest.Title))
                {
                    guest.Title = rand.Next(2) == 0 ? "Ms" : "Mr";
                }
                if (string.IsNullOrEmpty(guest.FirstName))
                {
                    guest.FirstName = char.ConvertFromUtf32(charAsInt);
                    charAsInt += 1; // names need to be unique
                }
                if (string.IsNullOrEmpty(guest.LastName))
                {
                    guest.LastName = "Test";
                }
            }
        }

        private async Task<string> GetNationalityAsync(PropertyDetails propertyDetails)
        {
            string nationality = string.Empty;
            if (!string.IsNullOrWhiteSpace(propertyDetails.ISONationalityCode))
            {
                nationality = await _support.TPNationalityLookupAsync(ThirdParties.STUBA, propertyDetails.ISONationalityCode);
            }
            if (string.IsNullOrEmpty(nationality))
            {
                nationality = _settings.LeadGuestNationality(propertyDetails);
            }
            return nationality;
        }

        private bool CostsAndCancellation(PropertyDetails propertyDetails, XmlDocument xml)
        {
            bool available = false;

            foreach (var room in propertyDetails.Rooms)
            {
                // Check costs
                var roomNode = xml.SelectSingleNode(string.Format("BookingCreateResult/Booking/HotelBooking/Room[RoomType/@code = '{0}']", room.RoomTypeCode));
                room.LocalCost = roomNode.SelectSingleNode("TotalSellingPrice/@amt").InnerText.ToSafeDecimal();
                room.GrossCost = roomNode.SelectSingleNode("TotalSellingPrice/@amt").InnerText.ToSafeDecimal();

                // Cancellation charges
                foreach (XmlNode cancelNode in roomNode.SelectNodes("CanxFees/Fee"))
                {
                    DateTime startDate;

                    if (cancelNode.SelectSingleNode("@from") is null)
                    {
                        startDate = DateTime.Now.Date;
                    }
                    else
                    {
                        startDate = cancelNode.SelectSingleNode("@from").InnerText.ToSafeDate();
                    }

                    var endDate = new DateTime(2099, 12, 25);

                    var nextStartDate = cancelNode.NextSibling;
                    if (nextStartDate is not null)
                    {
                        endDate = nextStartDate.SelectSingleNode("@from").InnerText.ToSafeDate().AddDays(-1);
                    }

                    decimal amount = cancelNode.SelectSingleNode("Amount/@amt").InnerText.ToSafeDecimal();
                    propertyDetails.Cancellations.AddNew(startDate, endDate, amount);
                }
            }

            propertyDetails.LocalCost = propertyDetails.Rooms.Sum(r => r.LocalCost);

            propertyDetails.Cancellations.Solidify(SolidifyType.Sum);
            available = xml.SelectNodes("BookingCreateResult/Booking/HotelBooking/Room").Count == propertyDetails.Rooms.Count;

            return available;
        }

        private void ExtractErrata(XmlDocument response, PropertyDetails propertyDetails)
        {
            var oErrataNodes = response.SelectNodes("BookingCreateResult/Booking/HotelBooking/Room/Messages/Message[Type = 'Supplier Notes']/Text");
            if (0 is var arg2 && (oErrataNodes?.Count) is { } arg1 && arg1 > arg2)
            {
                foreach (XmlNode oErratum in oErrataNodes)
                {
                    propertyDetails.Errata.AddNew("Important Informations", oErratum.InnerText);
                }
            }
        }

        public bool SupportsLiveCancellation(IThirdPartyAttributeSearch searchDetails, string source)
        {
            return _settings.AllowCancellations(searchDetails);
        }

        #region Stuff

        public bool SupportsRemarks => false;

        public bool SupportsBookingSearch => false;

        public int OffsetCancellationDays(IThirdPartyAttributeSearch searchDetails, string source)
        {
            return _settings.OffsetCancellationDays(searchDetails, false);
        }

        public bool RequiresVCard(VirtualCardInfo info, string source) => false;

        public ThirdPartyBookingSearchResults BookingSearch(BookingSearchDetails oBookingSearchDetails)
        {
            return new ThirdPartyBookingSearchResults();
        }

        public ThirdPartyBookingStatusUpdateResult BookingStatusUpdate(PropertyDetails oPropertyDetails)
        {
            return new ThirdPartyBookingStatusUpdateResult();
        }

        public string CreateReconciliationReference(string sInputReference)
        {
            return "";
        }

        public void EndSession(PropertyDetails oPropertyDetails)
        {
        }

        #endregion

    }
}